### ISSUE TITLE
Add support for control-flow and `Expr` values to `QuantumCircuit.depth`

### ISSUE THREAD

**Issue by jakelishman** [Qiskit/qiskit-terra#10229]
### What should we add?

In its current implementation, `QuantumCircuit.depth` iterates through the circuit, determining the maximum path length of the circuit without constructing a concrete `DAGCircuit`.  This path-length is dependent on classical wires as well as quantum ones, so the function would need to take `Expr` values in `.condition` and `.target` attributes into account.

Depends on #10223 and #10224.

**Comment by jakelishman** [comment]
Deferring this til later - `QuantumCircuit.depth` doesn't actually support control-flow at all right now.  We deferred added those in #8763 because it wasn't clear what the use-case or the API around them should be.  The same arguments still apply.
